source/shared/cpp/ObjectModel/MarkDownBlockParser.cpp

```diff
@@ -0,0 +1,417 @@
+{
+    /// MarkDown keywords
+    if (stream.peek() == '[' || stream.peek() == ']' || stream.peek() == ')' || stream.eof())
+    {
+        parser.Flush(token);
+        return Captured;
+    }
+
+    if (parser.IsMarkDownDelimiter(stream.peek()))
+    {
+        // encounterred first emphasis delimiter
+        parser.CaptureCurrentCollectedStringAsRegularToken();
+        DelimiterType emphasisType = EmphasisParser::GetDelimiterTypeForCharAtCurrentPosition(stream.peek()); 
+        parser.UpdateCurrentEmphasisRunState(emphasisType);
+        token += stream.get();
+        return Emphasis;
+    }
+    else
+    {    
+        parser.UpdateLookBehind(stream.peek());
+        token += stream.get();
+        return Text;
+    }
+}
+
+/// captures text untill it see none-emphasis character. When it does, switch to text state
+unsigned int EmphasisParser::MatchEmphasis(EmphasisParser &parser, std::stringstream &stream, std::string& token)
+{
+    if (stream.peek() == '[' || stream.peek() == ']' || stream.peek() == ')' || stream.eof())
+    {
+        parser.Flush(token);
+        return Captured;
+    }
+
+    if (parser.IsMarkDownDelimiter(stream.peek()))
+    {
+        DelimiterType emphasisType = EmphasisParser::GetDelimiterTypeForCharAtCurrentPosition(stream.peek());
+        if (parser.IsEmphasisDelimiterRun(emphasisType))
+        {
+            parser.UpdateCurrentEmphasisRunState(emphasisType);
+        }
+        token += stream.get();
+    }
+    else
+    {
+        if (stream.peek() != '\\')
+        {
+            parser.CaptureEmphasisToken(stream.peek(), token);
+        }
+        else
+        {
+            // skips escape char
+            parser.CaptureEmphasisToken(stream.peek(), token);
+            stream.get();
+        }
+
+        parser.ResetCurrentEmphasisState();
+        parser.UpdateLookBehind(stream.peek());
+        token += stream.get();
+
+        return Text;
+    }
+    return Emphasis;
+}
+
+// Captures remaining charaters in given token
+// and causes the emphasis parsing to terminate
+void EmphasisParser::Flush(std::string& currentToken)
+{
+    if (m_current_state != Text)
+    {
+        CaptureEmphasisToken(currentToken[0], currentToken);
+        m_delimiterCnts = 0;
+    }
+    else
+    {
+        CaptureCurrentCollectedStringAsRegularToken(currentToken);
+    }
+    currentToken.clear();
+}
+
+bool EmphasisParser::IsMarkDownDelimiter(char ch)
+{
+    return ((ch == '*' || ch == '_') && (m_lookBehind != Escape));
+}
+
+void EmphasisParser::CaptureCurrentCollectedStringAsRegularToken(std::string&currentToken) 
+{
+    if (currentToken.empty())
+        return;
+    std::shared_ptr<MarkDownStringHtmlGenerator> codeGen = 
+        std::make_shared<MarkDownStringHtmlGenerator>(currentToken);
+    std::shared_ptr<MarkDownHtmlGenerator> token = std::static_pointer_cast<MarkDownHtmlGenerator>(codeGen);
```

Comment:
You don't need to do this in two steps.  It can just be:
std::shared_ptr<MarkDownHtmlGenerator> token = std::make_shared<MarkDownStringHtmlGenerator>(currentToken);

Reply:
Thanks!

---
In reply to: [157608398](https://github.com/Microsoft/AdaptiveCards/pull/1045#discussion_r157608398) [](ancestors = 157608398)